### PR TITLE
feat: add support for returning all related reactions for a metabolite

### LINKED ISSUE #832: "See reactions from all compartments" button has no effect
**Describe the bug**
On metabolite pages in the GEM browser, selecting the `See reactions from all compartments` button doesn't seem to do anything.

For example, on the page for [`cholesterol[c]`](https://metabolicatlas.org/explore/Human-GEM/gem-browser/metabolite/MAM01450c), 25 reactions are shown to involve that metabolite. When `See reactions from all compartments` is selected, no additional reactions appear (still 25). However, if one looks at the page for [`cholesterol[l]`](https://metabolicatlas.org/explore/Human-GEM/gem-browser/metabolite/MAM01450l), 71 reactions are shown, suggesting that more reactions for `cholesterol` did exist.

**Steps to reproduce**
On the [`cholesterol[c]`](https://metabolicatlas.org/explore/Human-GEM/gem-browser/metabolite/MAM01450c) page, click on `See reactions from all compartments`.

**Expected behavior**
On a metabolite page, clicking the `See reactions from all compartments` button should show reactions involving any compartment form of that metabolite. In the specific example provided, since there are 71 reactions involving the `cholesterol[l]` metabolite, then clicking the `See reactions from all compartments` button on the `cholesterol[c]` page should increase the number of reactions shown to more than 71 (likely above 100).

### PR DESCRIPTION
**Pull request by e0**
**Related issue(s) and PR(s)**  
This PR closes #832 .

**Type of change**  

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

 
**List of changes made**  

- A new param called `allCompartments` is added to the related reactions for metabolite endpoint in The API.
- The frontend is adapted to send the new param, which results in different reactions.

**Screenshot of the fix**  

image

image

**Testing**  

- Relative urls that can be reused both for production and local testing
- Instructions on how to test

1. Visit `/explore/Human-GEM/gem-browser/metabolite/MAM01450c`.
2. Verify that there are 25 related reactions (as shown in the first screenshot).
3. Click the "See reactions from all compartments" button.
4. Verify that the related reactions table reloads and shows 193 reactions afterwards (as shown in the second screenshot).
5. Click the "Restrict to current compartment" button.
6. Verify that the related reactions table reloads and shows 25 reactions again.

**Checklist**  

- [x] My code follows the [NBIS style guidelines](https://github.com/NBISweden/development-guidelines)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have checked so that the same data as before is returned by the API

### CODE REVIEW

**Review by MalinAhlberg — APPROVED** [review]
Nice job!

The code looks good, more reactions are shown when clicking the button (tested different metabolites from different models) and, as far as I can tell, the reactions are correctly selected.

**Review by nanjiangshu — COMMENTED** [review]
Great work @e0. And it's nice to learn new ways of neo4j querying when reviewing your code. However, I left two comments regarding the variable name and the querying result.

File: api/src/endpoints/neo4j.js [hunk 1]
```diff
@@ -28,15 +28,34 @@ const neo4jRoutes = express.Router();
...
+      full,
+      searchTerm,
+    };
 
     if (componentTypes) {
       payload.componentTypes = JSON.parse(componentTypes);
     }
 
+    if (allCompartments) {
+      payload.allCompartments = JSON.parse(allCompartments);
```
[the remark this answers is not in the document]
Comment (e0) [reply api/src/endpoints/neo4j.js:56]: I like this suggestion! I didn't think about changing anything before as I just took the variable name we already had. I can add a commit for this.

File: api/src/neo4j/queries/relatedReactions.js [hunk 2]
```diff
@@ -32,8 +39,18 @@ WHERE ccms1 = ccms3 and r1.id <> r.id`;
...
     case NODE_TYPES.metabolite:
-      statement = `
+      // If `allCompartments` is true, all reactions associated with
+      // the metabolite are returned. Otherwise, only reactions associated
+      // with the same compartmentalized metabolite are returned.
+      if (allCompartments) {
+        statement = `
+MATCH (:CompartmentalizedMetabolite${m} {id: '${id}'})-[${v}]-(m:Metabolite)
+WITH m
+MATCH (m)-[${v}]-(:CompartmentalizedMetabolite${m})-[${v}]-(r:Reaction)`;
```
[the remark this answers is not in the document]
Comment (e0) [reply api/src/neo4j/queries/relatedReactions.js:49]: Ah that's a great catch! Looks like we are filtering away some of the relevant reactions in later parts of the query. Will try to fix this soon.

[the remark this answers is not in the document]
Comment (e0) [reply api/src/neo4j/queries/relatedReactions.js:49]: I see that this is actually just due to the limit we have, which is currently set to 200. We made some improvements in the load time a while ago and maybe it's ok to increase this limit now. I tested 1000 on the transport reactions subsystem page and it loaded in 15 seconds, which is acceptable I think. I will make this change for now and we can circle back to this later if needed.

File: api/src/endpoints/neo4j.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (e0) [reply api/src/endpoints/neo4j.js]: Fixed in https://github.com/MetabolicAtlas/MetabolicAtlas/pull/869/commits/8e2179dded21040ba5ade335a71b482586f58213

File: api/src/neo4j/queries/relatedReactions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (e0) [reply api/src/neo4j/queries/relatedReactions.js:49]: Fixed in: https://github.com/MetabolicAtlas/MetabolicAtlas/pull/869/commits/605d536cb6bb89a287ad602865c8b6f508843f03

**Comment by e0** [comment]
> Great work @e0. And it's nice to learn new ways of neo4j querying when reviewing your code. However, I left two comments regarding the variable name and the querying result.

Thanks for the careful review @nanjiangshu! Great that you spotted the issue with the number of reactions. I have added commits that address both of your points now.

**Review by nanjiangshu — APPROVED** [review]
Looks great now!

**Comment by mihai-sysbio** [comment]
It's great this was resolved in the current sprint 🌈

**Pull request merged**